### PR TITLE
Migrate to postcss-values-parser

### PR DESCRIPTION
**Pull request by edg2s**
> Which issue, if any, is this issue related to?

#2404

> Is there anything in the PR that needs further explanation?

Uses a more detailed parser that can handle Less/SCSS expressions, even if they have no spaces.

### CODE REVIEW

File: lib/rules/value-keyword-case/index.js [hunk 1]
```diff
@@ -65,14 +65,16 @@ const rule = function (expectation, options) {
         }
 
         // Ignore keywords within `url` and `var` function
-        if (node.type === "function" && (valueLowerCase === "url" || valueLowerCase === "var" || valueLowerCase === "counter" || valueLowerCase === "counters" || valueLowerCase === "attr")) {
-          return false
+        if (node.type === "func" && (valueLowerCase === "url" || valueLowerCase === "var" || valueLowerCase === "counter" || valueLowerCase === "counters" || valueLowerCase === "attr")) {
+          // Prevent further descending, but don't return false as that stops all walking
+          node.walk = null
```
Comment (hudochenkov) [inline lib/rules/value-keyword-case/index.js]: `node.walk` is a PostCSS API. I don't think it's a good idea, to break PostCSS API.

**Comment by jeddy3** [comment]
@edg2s Thanks for looking into this. 

The code itself looks great. However, I think this issue will need to trigger a side discussion about whether we wish to introduce another value parser _to support non-standard syntax_.

We're always striving to find the right balance between features and sustainable development. Endeavouring for as much consistency as possible across the rules has helped us keep things maintainable, and development sustainable.

I can think of three options going forward:

- Continue to only use `postcss-value-parser` and mark this bug as wontfix (consistency over features).
- Migrate wholesale (approx 30 files make use of it) to `postcss-values-parser` (consistent and features, but at an effort)
- Allow two value parsers and document when to use one over the other (features over consistency)

Related to https://github.com/stylelint/stylelint/issues/1183

**Comment by hudochenkov** [comment]
Also, we [might drop Less support](https://github.com/stylelint/stylelint/pull/1899) soon. Worth to consider before any changes related to Less support only.

**Comment by edg2s** [comment]
Currently the code has a whole bunch of rules for less/scss support which assume that the value is getting parsed correctly. I think there should either be support or no support (which would be a terrible shame as we'd probably have to stop using it at Wikimedia). The status quo is that we pretend the parser works when in fact it works 90% of the time.

**Comment by edg2s** [comment]
With regards to points 2 & 3, we could create a tech debt bug to migrate to values-parser and either do that in one go as you say, or piece by piece.

**Comment by edg2s** [comment]
Also the APIs were almost identical once I'd worked out how to do the same tree walk, so I don't think such a migration would be that hard, and would be happy to help.

**Comment by evilebottnawi** [comment]
@edg2s What benefits do we get besides supporting less? Speed, more correct ast and etc? Then it will be easier to decide whether it is necessary or not.

**Comment by edg2s** [comment]
Someone else may have a better test rig for determining performance, although as a more sophisticated parser it may end up being slightly slower. Certainly it appears to be better at parsing expressions so it's conceivable it might be useful for vanilla CSS, such as the [calc() function](https://developer.mozilla.org/en/docs/Web/CSS/calc).

**Comment by jeddy3** [comment]
@edg2s Thanks for sharing your further thoughts.

> I think there should either be support or no support (which would be a terrible shame as we'd probably have to stop using it at Wikimedia).

Is there anyone at Wikimedia with experience with parsers (or is wanting to learn) who can pick up the maintenance of `postcss-less`?

> Someone else may have a better test rig for determining performance

You can test the performance difference of your changes to the `value-keyword-case` rule using stylelint's [benchmarking tool](https://stylelint.io/developer-guide/rules/#improving-the-performance-of-a-new-or-an-existing-rule). Just run it against `master` and then again against your branch, and then post your results below.

> Certainly it appears to be better at parsing expressions so it's conceivable it might be useful for vanilla CSS, such as the calc() function.

That's a good point. There's a [summary of other differences](https://github.com/lesshint/postcss-values-parser#postcss-values-parser-vs-postcss-value-parser) within the `postcss-values-parser` README.

> Also the APIs were almost identical once I'd worked out how to do the same tree walk, so I don't think such a migration would be that hard, and would be happy to help.

Thanks for the offer!

If @edg2s is willing to do the legwork, and if the performance difference is acceptable given the new features, I think option 2 is a valid route to go down. It sounds like the behaviour of the parser is more similar to the ones in PostCSS and `postcss-selector-parser`, and so we might be better positioned if value and selector parsing are rolled into PostCSS in the future. Also, I think the consistency will help new contributors maintain the codebase.

**Comment by davidtheclark** [comment]
I agree with @jeddy3. 👍

**Comment by jeddy3** [comment]
>  and if the performance difference is acceptable given the new features,

I ran the `value-keyword-case` rule of `master` and @edg2s fork through the benchmarking script and there appears to be no discernible difference in performance. As such, I think option 2 is good to go.

@edg2s Is this something you'd be willing to champion in this PR?

**Comment by edg2s** [comment]
> Is there anyone at Wikimedia with experience with parsers (or is wanting to learn) who can pick up the maintenance of postcss-less?
Well MediaWiki uses its own markup language so in many teams it is a required skill :) I can certainly ask around.

I'll have to come back to converting the other 30 as it will still take a fair bit more time than this patch, but if anyone else wants to get started on it...

**Comment by jeddy3** [comment]
> I can certainly ask around.

That'd be awesome thanks.

> I'll have to come back to converting the other 30 as it will still take a fair bit more time than this patch, but if anyone else wants to get started on it...

No worries, we'll leave this PR open for now so you can work on it when you have time. I'll also create a new issue to see if anyone else is able to chip away at it in the meantime.

**Comment by edg2s** [comment]
So working on a few new rules - it appears that values-parser has some issues of it's own, mostly with SASS (as it was written with LESS in mind). I filed a few issues, but the maintainer was sadly less than helpful. Given than both parsers have their limitations with extended syntax the only option at this stage would be to use whichever one is best for the job.

**Comment by davidtheclark** [comment]
@edg2s, thanks for investigating. Too bad it didn't work out. Seems the smoothest path right now would be to stick with the value parser we have and add exceptions and workarounds as needed.

**Comment by jeddy3** [comment]
@edg2s Yes, thank you! From reading those `values-parser` issues, it looks like you dived deeply into making this PR work.

> Seems the smoothest path right now would be to stick with the value parser we have and add exceptions and workarounds as needed.

What that in mind, should this PR now be merged? If so, I'd like to see a brief comment in the source code that highlights the use of `values-parser` in the rules and what it works around.

**Comment by evilebottnawi** [comment]
`postcss-values-parser` works as expected, because all mathematical expression should be interpreted as they inside in `calc` function (`-` and `+` always have space before and after sign). I.e `10+@value` and `10-@value` should be interpreted as is, only one thing don't work as expected `10*@value` and `10/@value`. `-@value` and `+@value` should be interpreted as is,
`*@value` and `/@value` should be throw error  (need create issue and send PR).

My opinion on this issue:
Creating a new parser(`postcss-values-parser`) was a bad idea, you could come to an agreement and create one handy parser for values, guys from `postcss` responsive developers. Now we have two parsers with weird and incorrect behavior. 

Nothing complicated, we just need to better interact with each other. 
I apologize if this sounded a little rude.

BTW, if fix this behavior in `stylelint`

**Comment by jeddy3** [comment]
Closing in favour of https://github.com/stylelint/stylelint/pull/2558

Thanks for your work on this, but another way was found that uses the existing parser.

**Pull request closed without merging**